core/shared/src/main/scala/scalaz/zio/STM.scala

```diff
@@ -0,0 +1,226 @@
+            }
+            (journal, leftMapped.swap)
+          }
+      }
+    )
+
+  /**
+   * Maps the value produced by the computation.
+   */
+  def map[B](f: A => B): STM[E, B] =
+    STM(
+      journal =>
+        self.run(journal) match {
+          case (journal, either) => {
+            val eitherMapped = either match {
+              case Right(a) => Right(f(a))
+              case Left(b)  => Left(b)
+            }
+            (journal, eitherMapped)
+          }
+      }
+    )
+
+  /**
+   * Feeds the value produced by this computation to the specified computation,
+   * and then runs that computation as well to produce its failure or value.
+   */
+  def flatMap[E1 >: E, B](f: A => STM[E1, B]): STM[E1, B] =
+    STM(
+      journal =>
+        self.run(journal) match {
+          case (journal, Left(error)) => (journal, Left(error))
+          case (journal, Right(a))    => f(a).run(journal)
+      }
+    )
+
+  /**
+   * Tries this computation first, and if it fails, tries the other computation.
+   */
+  def orElse[E2, A1 >: A](that: => STM[E2, A1]): STM[E2, A1] =
+    self.attempt.flatMap(either => either.fold(_ => that, STM.point(_)))
+
+  /**
+   * Surfaces any errors at the value level, where they can be recovered from.
+   */
+  def attempt: STM[Nothing, Either[E, A]] =
+    STM(
+      journal =>
+        self.run(journal) match {
+          case (journal, value) => (journal, Right(value))
+      }
+    )
+
+  /**
+   * Sequentially zips this value with the specified one.
+   */
+  def seq[E1 >: E, B](that: => STM[E1, B]): STM[E1, (A, B)] =
+    self.flatMap(a => that.map(b => (a, b)))
+}
+
+object STM {
+
+  object internal {
+    def freshIdentity(): Long = counter.incrementAndGet()
+
+    private val counter: AtomicLong = new AtomicLong()
+
+    val semaphore = new java.util.concurrent.Semaphore(1)
+
+    abstract class Entry {
+      type A
+      val tVar: TVar[A]
+      val newValue: A
+      val expectedVersion: Long
+
+      def isValid: Boolean = tVar.version == expectedVersion
+      def commit(): Unit   = tVar.value = newValue
+    }
+
+    object Entry {
+
+      def apply[A0](tVar0: TVar[A0], newValue0: A0, expectedVersion0: Long): Entry = new Entry {
+        type A = A0
+        val tVar            = tVar0
+        val newValue        = newValue0
+        val expectedVersion = expectedVersion0
+      }
+    }
+  }
+
+  import internal._
+
+  /**
+   * A variable that can be modified as part of a transactional computation.
+   */
+  class TVar[A] private (identity: Long, @volatile var version: Long, @volatile var value: A) {
+    self =>
+
+    final def get: STM[Nothing, A] =
+      STM(journal => {
+        val entry = journal.getOrElse(identity, Entry(self, value, version))
+
+        (journal.updated(identity, entry), Right(value))
+      })
+
+    final def set(newValue: A): STM[Nothing, Unit] =
+      STM(journal => {
+        val entry = journal.getOrElse(identity, Entry(self, newValue, version))
```

Comment:
Isn't newValue lost if an entry was already in the journal?  We need to read the journal, because if there was an older read with an older version, we must keep that version. But we must also keep the newValue. Maybe :
val entry = Entry(self, newValue, journal.get(identity).fold(e => e.version, version))
Also, if an STM just writes without having read it before, shouldn't the entry be considered always valid? That would mean "no version", either with an Option or with a special value.

Reply:
You are right about this. The new value must be preserved.